### PR TITLE
Allow constructing subclass of abstract class

### LINKED ISSUE #774: autocxx does not recognize concrete subclasses of abstract base classes
## Minimal Working Example

(I _think_ it's minimal. I'll try to add a pull request with a new test in due course.)

Test.h
```
#pragma once

namespace test {
  class A;
  class B;
}

namespace test {

class A {
public:
  A() {}
  ~A() {}
  // the following line makes Test2 Opaque, and deprives it of a make_unique()
  // comment next line, uncomment the one after to obtain a make_unique() for Test2
  virtual int b() = 0;
  // int b() { return 2; }
};
  
class B: public A {
public:
  B() {}
  ~B() {}
  int b() { return 3; }
};

}
```

build.rs
```
fn main() {
    let path = std::path::PathBuf::from("src");
    let mut b = autocxx_build::Builder::new("src/main.rs", &[&path]).expect_build();
    b.flag_if_supported("-std=c++14").compile("autocxx-demo");
    println!("cargo:rerun-if-changed=src/main.rs");
}
```

main.rs
```
use autocxx::prelude::*;

include_cpp! {
    #include "Test.h"
    safety!(unsafe)
    generate!("test::B")
}

fn main() {
    // I'd like to initialize a B here, but I can't
}
```

## Expected Behavior

The generated `struct B` has a `make_unique` implementation.

## Actual Behavior

The generated `struct B` has a `Drop` implementation but no `make_unique`. It is not clear how I could instantiate a variable of type `B`.

## Steps to Reproduce the Problem

See above

## Specifications

- Version: 0.16.0 (or later if there's a later)
- Platform: Windows

### PR DESCRIPTION
**Pull request by russelltg**
Fixes #774

I have a solid start here, but I think I'm going to need some guidance to get this complete. I have it all working except overloads, and I need to somehow get the C++ types of the arguments from an `Api<FnPrePhase2>`

I put in a placeholder `args: Vec<QualifiedName>` in my `Signature` object but I'm not seeing anything in `Function` that I could use to fill that. Also, ideally the arguments wouldn't be the QualifiedName so type aliases don't get it confused

- [ ] Tests pass
- [ ] Appropriate changes to README are included in PR

### CODE REVIEW

**Comment by adetaylor** [comment]
Thanks for having a stab at this! Looks to be along the right lines, and good job figuring out the convoluted codebase well enough to get this in the right places.

> I need to somehow get the C++ types of the arguments from an Api<FnPrePhase2>

`FnAnalysis::params` is what you need. This contains the Rust types, but they're "converted", which means we have things like `UniquePtr<Foo>` rather than whatever bindgen gave us. Therefore the `syn::Type`s in that field are essentially canonical and can be relied upon to uniquely represent C++ types; they should be good enough for overload resolution. 
 i.e. Rather than compare `QualifiedName`s, you should compare `syn::Type`s, which is what you'll find in this location.

I'm not going to take a more detailed look at this until the tests are green and you tell me that it's good to go. Good luck!

**Comment by russelltg** [comment]
Thanks for the help there, I've figured out the resolution stuff. 

The only missing thing is it doesn't really support pure virtual dtors in either the base or the sub class. I don't think pure virtual dtors get passed to the analysis? Or at least I don't see it.

**Comment by russelltg** [comment]
I'm going to go ahead and mark this as ready for review, as I think it's useful and correct as is. I'm happy to deal with the pure virtual dtor edgecase if you think that's important/worthwhile.

**Comment by adetaylor** [comment]
Thanks, I'm travelling right now so I'm unlikely to take a look at this for a few days, but I'll get there.

File: integration-tests/tests/integration_test.rs [hunk 1]
```diff
@@ -11766,6 +11766,112 @@ fn test_issue_1170() {
...
+            void b(int) { }
+            void b(bool) = 0;
+
+            int c() const { return 1; }
+            int c() { return 2; }
+        };
+
+        // class Partial5 : public Base {
+        // public:
+        //     ~Partial5() = 0;
```
Comment (adetaylor) [inline integration-tests/tests/integration_test.rs:11838]: Totally happy not to address this case now, but please could you add a comment in this test case explaining why this is commented out.

Comment (adetaylor) [reply integration-tests/tests/integration_test.rs:11838]: (also if you could file a new github issue that'd be great)

File: integration-tests/tests/integration_test.rs [hunk 2]
```diff
@@ -11766,6 +11766,112 @@ fn test_issue_1170() {
...
+        //     void b(int) { }
+        //     void b(bool) { }
+
+        //     int c() const { return 1; }
+        //     int c() { return 2; }
+        // };
+
+    "};
+    let rs = quote! {
+        static_assertions::assert_impl_all!(ffi::FullyDefined: moveit::CopyNew);
```
[the remark this answers is not in the document]
Comment (adetaylor) [reply integration-tests/tests/integration_test.rs:11851]: I wish I'd thought of using this for loads of the other tests!

**Review by adetaylor — CHANGES_REQUESTED** [review]
Sorry for the delay on reviewing this.

It's looking great!

My comments are:
* a load of nit picks and suggestions
* one substantial comment that I can't easily follow what happens in cases where our data structures are incomplete (for unforeseen reasons) - I want to make sure we're erring on the side of regarding classes as abstract. I'm sure I could figure it out by peering more carefully at it, but the fact it's hard to follow suggests it could do with a little bit more in the way of comments.

**Comment by russelltg** [comment]
`fields_and_bases_first` is super cool, thanks for pointing me there, that *significnatly* simplifies the code. The only problem....it doesn't seem to actually do bases first? It's definitely emitting the `Struct` api's before the methods.....

I suppose the name is evident--it's fields and bases first not fields, methods, and bases first. I think the cleanest way to fix this would be to either have a new traverser that does to methods first (or just edit fields_and_bases_first to do that) or just go through the API list twice like I was doing previously

**Comment by adetaylor** [comment]
> I suppose the name is evident--it's fields and bases first not fields, methods, and bases first. I think the cleanest way to fix this would be to either have a new traverser that does to methods first (or just edit fields_and_bases_first to do that)

Originally it did do this, but it turned out this caused cycles, because (I think) some types of constructor need to depend on the type they're constructing. You'd be most welcome to try to solve this but I think it was Hard. (I can't remember the details). So going through the list twice is probably best.

File: engine/src/conversion/apivec.rs [hunk 3]
```diff
@@ -97,6 +97,10 @@ impl<P: AnalysisPhase> ApiVec<P> {
         self.apis.iter()
     }
 
+    pub(crate) fn iter_mut(&mut self) -> impl Iterator<Item = &mut Api<P>> {
+        self.apis.iter_mut()
```
Comment (adetaylor) [inline engine/src/conversion/apivec.rs:101]: Oh, I'm really sorry, I offered you bad advice here. `iter_mut` is considered harmful here for the reasons explained in this comment:
https://github.com/google/autocxx/pull/1206/files/8d4d785f4105c3332b1f77f8f0c0756ef9e72545..c90cd896eca376977c87e04707ec48ba01b243bd#diff-dd61892ff96d055fbcab83c5eecc76c8b6120c3eb2f1b5687a8c025435200fcdL32
I just plain forgot! Sorry for the rework.

File: engine/src/conversion/analysis/abstract_types.rs [hunk 4]
```diff
@@ -6,115 +6,167 @@
...
+    #[derive(Default, Debug, Clone)]
+    struct ClassAbstractState {
+        undefined: HashSet<Signature>,
+        defined: HashSet<Signature>,
+    }
+    let mut class_states: HashMap<QualifiedName, ClassAbstractState> = HashMap::new();
+    let mut abstract_classes = HashSet::new();
+
+    for api in fields_and_bases_first(apis.iter()) {
+        match dbg!(api) {
```
Comment (adetaylor) [inline engine/src/conversion/analysis/abstract_types.rs:73]: Don't forget to remove this dbg

**Comment by russelltg** [comment]
This is ready for review @adetaylor  :)

**Comment by adetaylor** [comment]
I'm good to merge this once the formatting problems are fixed!

**Comment by adetaylor** [comment]
Thanks very much for this!
